src/python/pants/reporting/json_reporter.py

```diff
@@ -0,0 +1,144 @@
+# coding=utf-8
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import json
+import os
+from builtins import list, open, str
+from collections import defaultdict, namedtuple
+
+from future.moves.itertools import zip_longest
+from six import string_types
+
+from pants.base.build_environment import get_buildroot
+from pants.reporting.reporter import Reporter
+from pants.util.dirutil import safe_mkdir
+
+
+class JsonReporter(Reporter):
+  """A reporter to capture workunit data into a JSON structure."""
+
+  _log_level_str = [
+    'FATAL',
+    'ERROR',
+    'WARN',
+    'INFO',
+    'DEBUG',
+  ]
+
+  # JSON reporting settings.
+  #   json_dir: Where the report file goes.
+  Settings = namedtuple('Settings', Reporter.Settings._fields + ('json_dir',))
+
+  def __init__(self, run_tracker, settings):
+    super(JsonReporter, self).__init__(run_tracker, settings)
+    # The main report, and associated tool outputs, go under this dir.
+    self._json_dir = settings.json_dir
+
+    self._buildroot = get_buildroot()
+    self._report_file = None
+
+    # Results of the state of the build.
+    self._results = {}
+    # We use a stack to track the nested workunit traversal of each root
```

Comment:
Nit: Period. 

This matters (even though only a tiny bit) because finished sentences imply finished thoughts. I don't want readers thinking there's an "and" they're missing...